[GLib] Add introspectable API to configure experimental features at runtime

aperezdc opened the pull request:
#### 760e6a2591cf8cb0e8bc40af0caaccce3bc64f28
<pre>
[GLib] Add introspectable API to configure experimental features at runtime
<a href="https://bugs.webkit.org/show_bug.cgi?id=255518">https://bugs.webkit.org/show_bug.cgi?id=255518</a>

Reviewed by Carlos Garcia Campos.

Adds a WebKitFeature boxed type that wraps API::Feature, and new
functions in WebKitSettings to change their values and retrieve the
lists of available features. The structure loosely follows what the
Cocoa port does, with a few notable differences:

- Some feature identifiers have an &quot;Enabled&quot; suffix, and others do
  not, as declared in UnifiedWebPreferences.yaml. The identifiers
  exposed though the API remove the suffix for consistency.
- Whether a feature is hidden is not exposed. There is currently only
  one hidden feature (fullscreen API) so it seems moot to make the
  distinction. Might be added in the future if needed.
- The feature category is a string, to avoid committing a set of fixed
  enum values to the public API. This way it will not be needed to
  map values if there are changes in the future, and also transmits the
  idea that the category for a given feature is informative, and may
  change across versions.
- There is a helper WebKitFeatureList boxed type that groups features,
  to avoid using GPtrArray (which is bad for GObject Introspection
  bindings) or GList (poor match given the sizes of each set are fixed
  and well known).

The GTK MiniBrowser gets two new pages in its settings dialog, one for
experimental features and the other for development features. This
needed a switcher added, so at the same time the dialog gets changed to
use a header bar where to place a stack switcher. The existing page
with the settings gets tweaked to allow type-to-search, as it was done
also for the two new pages.

Both the GTK and WPE MiniBrowser get a new -F/--features= command line
option which can be used to toggle a set of features given their
identifiers.

* Source/WebKit/PlatformGTK.cmake: Add WebKitFeature.h.in to the list of
  generated API headers.
* Source/WebKit/PlatformWPE.cmake: Ditto.
* Source/WebKit/SourcesGTK.txt: Add WebKitFeature.cpp to the list of
  sources.
* Source/WebKit/SourcesWPE.txt: Ditto.
* Source/WebKit/UIProcess/API/glib/WebKitFeature.cpp: Added.
(toFeatureStatus):
(toFeatureCategory):
(_WebKitFeature::_WebKitFeature):
(_WebKitFeature::toIdentifier):
(webkitFeatureCreate):
(webkitFeatureGetFeature):
(webkit_feature_ref):
(webkit_feature_unref):
(webkit_feature_get_identifier):
(webkit_feature_get_name):
(webkit_feature_get_details):
(webkit_feature_get_status):
(webkit_feature_get_category):
(webkit_feature_get_default_value):
(_WebKitFeatureList::_WebKitFeatureList):
(_WebKitFeatureList::~_WebKitFeatureList):
(webkitFeatureListCreate):
(webkit_feature_list_ref):
(webkit_feature_list_unref):
(webkit_feature_list_get_length):
(webkit_feature_list_get):
* Source/WebKit/UIProcess/API/glib/WebKitFeature.h.in: Added.
* Source/WebKit/UIProcess/API/glib/WebKitFeaturePrivate.h: Added.
* Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
(webkit_settings_set_feature_enabled):
(webkit_settings_get_feature_enabled):
(webkit_settings_get_all_features):
(webkit_settings_get_experimental_features):
(webkit_settings_get_development_features):
* Source/WebKit/UIProcess/API/glib/WebKitSettings.h.in:
* Source/WebKit/UIProcess/API/glib/webkit.h.in:
* Tools/MiniBrowser/gtk/BrowserSettingsDialog.c:
(featureTreeViewRowActivated):
(cellRendererToggled):
(featureTreeViewRenderEnabledData):
(featureTreeViewRenderStatusData):
(featureTreeViewRenderCategoryData):
(featureTreeViewSearchMatchSubstring):
(createFeatureTreeView):
(settingsTreeViewSearchMatchSubstring):
(browser_settings_dialog_init):
(browserSettingsDialogConstructed):
(browser_settings_dialog_new):
* Tools/MiniBrowser/gtk/main.c:
(findFeature):
(parseFeaturesOptionCallback):
(addSettingsGroupToContext):
* Tools/MiniBrowser/wpe/main.cpp:
(findFeature):
(activate):
(main):
* Tools/Scripts/webkitpy/style/checker.py:
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitSettings.cpp:
(testWebKitFeatures):
(beforeAll):

Canonical link: <a href="https://commits.webkit.org/263331@main">https://commits.webkit.org/263331@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/45366a445d25047a28a67d25427c997bf90111bf

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/4218 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/4339 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/4457 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/5688 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/4464 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/4211 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/4446 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/4304 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/4683 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/4279 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/4456 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3804 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/5682 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/4272 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/1951 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/3781 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/5963 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/3778 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/3847 "3 failures") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/5373 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/4258 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/3455 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/3768 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3779 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1049 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/7861 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/4112 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitFeature.cpp:
```diff
@@ -0,0 +1,150 @@
+    return toFeatureCategory(feature->feature->category());
+}
+
+gboolean webkit_feature_get_default(WebKitFeature* feature)
```
get_default_value?

aperezdc replied:
That does sound a bit clearer, I'll apply the change before pushing again 👌🏼

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitFeature.cpp:
```diff
@@ -0,0 +1,150 @@
+    return feature->feature->defaultValue() ? TRUE : FALSE;
+}
+
+gboolean webkit_feture_is_hidden(WebKitFeature* feature)
```
What's a hidden feature? should we expose hidden features in this API?

aperezdc replied:
It's the features with `hidden: true` in `UnifiedWebPreferences.yaml`... currently only the Fullscreen API feature may be hidden depending on a preprocessor symbol at build time, and it is not among the list of experimental features. Therefore I am leaning towards removing this function for now, and revisiting the topic later on if we see the need.

mcatanzaro replied:
I'd remove it, yes.

carlosgcampos reviewed (commented):
I like it, it's simpler than I expected

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitFeature.h.in:
```diff
@@ -0,0 +1,115 @@
+    WEBKIT_FEATURE_CATEGORY_NETWORK,
+    WEBKIT_FEATURE_CATEGORY_PRIVACY,
+    WEBKIT_FEATURE_CATEGORY_SECURITY,
+} WebKitFeatureCategory;
```
I'm nervous about these two because I expect feature status and feature category enums to change in the future. Understanding that exposing them makes the API more powerful, I'm not sure it's worth the risk, unless there is corresponding public Cocoa API. You can do most of what you want to do without these, right? The features list just would not be sorted so nicely?

aperezdc replied:
Apple ports have [private/unstable API](https://github.com/WebKit/WebKit/blob/0efe0345caa687a0ac9a5e914ce96b9de1ce29da/Source/WebKit/UIProcess/API/Cocoa/_WKFeature.h#L35) for features, and both enumerations are exposed.

The feature status is quite informative and WebKit itself uses it internally to decide some defaults, and what should be considered “experimental and can be shown in the Safari TP ‘Develop’ menu”; or otherwise “definitely internal or very much in development and applications shouldn't show this”. So at least I would like to keep `WebKitFeatureStatus`.

Categories are nice to group features together (there are a lot of them!) so I was thinking of doing something like the following, but replacing month names with categories:

![Screenshot from 2023-04-18 23-17-22](https://user-images.githubusercontent.com/723451/232895249-0aed658b-25ac-46fd-8651-5484f9c47edf.png)

If we are wary about the values of the `WebKitFeatureCategory` changing between releases, we could make the category a string, and then we would be free to return different strings as releases go by. It would be a bit annoying to not know in advance how many categories are there before enumerating the features, but it would be more flexible in exchange for making the API a bit less convenient. Thoughts?

mcatanzaro replied:
Private/unstable is the thing... maybe find the relevant Apple developer and talk to them about stability before making this all public API?

We can try to reduce the risk of breakage in the future:

> The feature status is quite informative and WebKit itself uses it internally to decide some defaults, and what should be considered “experimental and can be shown in the Safari TP ‘Develop’ menu”; or otherwise “definitely internal or very much in development and applications shouldn't show this”. So at least I would like to keep WebKitFeatureStatus.

There are so many status types... really a lot. I would fear these could be consolidated in the future. Realistically, there are so many states and not amazing distinctions between them (e.g. "stable" vs. "mature") that they feel ripe for consolidation. At least externally, if not also internally.

We can at least try to simplify somewhat at the public layer. I'd say UNSTABLE, DEVELOPER, TESTABLE, and PREVIEW can all be compressed together. STABLE and MATURE can be compressed. That leaves EMBEDDER (not quite sure about the intended semantics of this) and INTERNAL (what is this? I guess we should just hide features that are INTERNAL?). The fewer statuses we expose, the lower the probability that we have trouble and have to deprecate statuses in the future.  

> If we are wary about the values of the WebKitFeatureCategory changing between releases, we could make the category a string, and then we would be free to return different strings as releases go by. It would be a bit annoying to not know in advance how many categories are there before enumerating the features, but it would be more flexible in exchange for making the API a bit less convenient. Thoughts?

Problem solved. I was going to suggest exactly that.

mcatanzaro replied:
> PREVIEW

I suppose this status could allow us to enable certain features in Tech Preview while leaving them disabled elsewhere, so maybe it could be useful... but we can decide that later. It's easy to change our mind and expose a new category in the future. But we can't get rid of a status once we expose it; the best we can do is deprecate and just not have any features associated with that status anymore.

aperezdc replied:
Please check the updated patch, which has most of the documentation and includes descriptions for all the `WebKitFeatureStatus` enum values, and let me know if you still think they are unclear and/or some should be merged. To me all statuses look useful.

And indeed, we can use the status to decide which features to enable by default in Ephy TP, and which ones to show in some UI there to allow toggling. Note that `webkit_settings_get_{experimental,development}_features()` are convenience functions that return feature lists filtered by the status 😉. So what Ephy TP most likely would do is pick the experimental list, and enable “preview” features by default, and allow toggling all of “testable”, “preview”, and “stable” (maybe even “developer”).

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.h.in:
```diff
@@ -540,6 +540,21 @@ webkit_settings_get_disable_web_security                       (WebKitSettings *
+webkit_settings_get_feature_enabled                            (WebKitSettings *settings,
+                                                                WebKitFeature  *feature);
+
+WEBKIT_API GPtrArray *
```
[This one needs a rethink](https://www.bassi.io/articles/2023/02/20/bindable-api-2023/)

"Whatever you do, however low you feel on the day, whatever particular tragedy befell your family at some point, please: never use GLib array types in your API. Nothing good will ever come of it, and you’ll just spend your days regretting this choice."

Better options:

 * Return a WebKitFeature** and warn that each element needs to be deleted with webkit_feature_unref() then the container deleted with g_free() (um, assuming it's possible to express this via introspection annotations... I'm not sure if it is)
 * Just create a WebKitFeatureList type

aperezdc replied:
Ahhh, that's why I was getting the back of my head a bit itchy while writing `GPtrArray`... I think we can have a new `WebKitFeatureList` the same way we have `WebKitBackForwardList` and GLib has `GUnixFDList`.

(While on topic, it looks like we have a few functions return `GList`, like `webkit_context_menu_get_items`, `webkit_cookie_manager_get_cookies_finish`, etc. and maybe it was an oversight to let them be while working on the 2022 API 😨)

mcatanzaro reviewed (commented):
Concept looks good. This will be useful.

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitFeature.h.in:
```diff
@@ -0,0 +1,131 @@
+
+/**
+ * WebKitFeatureStatus:
+ * @WEBKIT_FEATURE_STATUS_EMBEDDER: Feature that adjust behaviour for
+ *   specific application needs. The feature is not part of a Web platform
+ *   feature, not a mature feature intended to be always on.
```
These sound like things that should be exposed via WebKitSettings?

mcatanzaro replied:
I wonder if we could find a different name for this, although I don't have any grand proposals.

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitFeature.h.in:
```diff
@@ -0,0 +1,131 @@
+ * @WEBKIT_FEATURE_STATUS_EMBEDDER: Feature that adjust behaviour for
+ *   specific application needs. The feature is not part of a Web platform
+ *   feature, not a mature feature intended to be always on.
+ * @WEBKIT_FEATURE_STATUS_UNSTABLE: Feature in development. The feature
+ *   may be unfinished, and there are no guarantees about its safety and
+ *   stability.
```
UNSTABLE sounds very similar to TESTABLE, and there's not much gap between that and PREVIEW. I would merge them.

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitFeature.h.in:
```diff
@@ -0,0 +1,131 @@
+ * @WEBKIT_FEATURE_STATUS_UNSTABLE: Feature in development. The feature
+ *   may be unfinished, and there are no guarantees about its safety and
+ *   stability.
+ * @WEBKIT_FEATURE_STATUS_INTERNAL: Feature for debugging the WebKit engine.
+ *   The feature is not generally useful for user or web developers, and
+ *   always disabled by default.
```
I don't understand why a feature would use this status. Maybe we can just ignore these?

aperezdc replied:
Why would we? They can be useful for us folks working in WebKit itself. There is zero cost in having one enum value here.

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitFeature.h.in:
```diff
@@ -0,0 +1,131 @@
+ * @WEBKIT_FEATURE_STATUS_INTERNAL: Feature for debugging the WebKit engine.
+ *   The feature is not generally useful for user or web developers, and
+ *   always disabled by default.
+ * @WEBKIT_FEATURE_STATUS_DEVELOPER: Feature for web developers. The feature
+ *   is not generally useful for end users, and always disabled by default.
```
So these are _web developer_ features, not WebKit developer features? It's news to me. :P 

What's the point of having features for developers that are always disabled? The only way to turn them on is to use Safari Tech Preview, correct? Or does the stable Safari allow toggling individual features?

aperezdc replied:
Safari stable has a hidden “Develop” menu which can be enabled, and features listed there individually. IIUC in Safari TP the menu is always shown.

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitFeature.h.in:
```diff
@@ -0,0 +1,131 @@
+ * @WEBKIT_FEATURE_STATUS_STABLE: Feature ready for general use. The
+ *   feature is enabled by default, but it may still be toggled to support
+ *   debugging and testing.
+ * @WEBKIT_FEATURE_STATUS_MATURE: Feature in general use. The feature is
+ *   always enabled and in general there should be no user-facing interface
+ *   to toggle it.
```
If there should be no UI, then there's no point in exposing this feature status.

aperezdc replied:
Note the *should*, not *must* 😉

mcatanzaro reviewed (changes requested):
Using strings for the category feels a lot more comfortable.

I'm still concerned about the feature status enum. There are too many statuses with small distinctions between them.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitFeature.cpp:
```diff
@@ -0,0 +1,431 @@
+ *
+ * The default value may be used by applications which include user interface
+ * to toggle features to restore its settings to their defaults. Note that
+ * the whether a feature is actually enabled must be checked with
```
```suggestion
 * whether a feature is actually enabled must be checked with
```

aperezdc committed:
[GLib] Add introspectable API to configure experimental features at runtime
https://bugs.webkit.org/show_bug.cgi?id=255518

Reviewed by Carlos Garcia Campos.

Adds a WebKitFeature boxed type that wraps API::Feature, and new
functions in WebKitSettings to change their values and retrieve the
lists of available features. The structure loosely follows what the
Cocoa port does, with a few notable differences:

- Some feature identifiers have an "Enabled" suffix, and others do
  not, as declared in UnifiedWebPreferences.yaml. The identifiers
  exposed though the API remove the suffix for consistency.
- Whether a feature is hidden is not exposed. There is currently only
  one hidden feature (fullscreen API) so it seems moot to make the
  distinction. Might be added in the future if needed.
- The feature category is a string, to avoid committing a set of fixed
  enum values to the public API. This way it will not be needed to
  map values if there are changes in the future, and also transmits the
  idea that the category for a given feature is informative, and may
  change across versions.
- There is a helper WebKitFeatureList boxed type that groups features,
  to avoid using GPtrArray (which is bad for GObject Introspection
  bindings) or GList (poor match given the sizes of each set are fixed
  and well known).

The GTK MiniBrowser gets two new pages in its settings dialog, one for
experimental features and the other for development features. This
needed a switcher added, so at the same time the dialog gets changed to
use a header bar where to place a stack switcher. The existing page
with the settings gets tweaked to allow type-to-search, as it was done
also for the two new pages.

Both the GTK and WPE MiniBrowser get a new -F/--features= command line
option which can be used to toggle a set of features given their
identifiers.

* Source/WebKit/PlatformGTK.cmake: Add WebKitFeature.h.in to the list of
  generated API headers.
* Source/WebKit/PlatformWPE.cmake: Ditto.
* Source/WebKit/SourcesGTK.txt: Add WebKitFeature.cpp to the list of
  sources.
* Source/WebKit/SourcesWPE.txt: Ditto.
* Source/WebKit/UIProcess/API/glib/WebKitFeature.cpp: Added.
(toFeatureStatus):
(toFeatureCategory):
(_WebKitFeature::_WebKitFeature):
(_WebKitFeature::toIdentifier):
(webkitFeatureCreate):
(webkitFeatureGetFeature):
(webkit_feature_ref):
(webkit_feature_unref):
(webkit_feature_get_identifier):
(webkit_feature_get_name):
(webkit_feature_get_details):
(webkit_feature_get_status):
(webkit_feature_get_category):
(webkit_feature_get_default_value):
(_WebKitFeatureList::_WebKitFeatureList):
(_WebKitFeatureList::~_WebKitFeatureList):
(webkitFeatureListCreate):
(webkit_feature_list_ref):
(webkit_feature_list_unref):
(webkit_feature_list_get_length):
(webkit_feature_list_get):
* Source/WebKit/UIProcess/API/glib/WebKitFeature.h.in: Added.
* Source/WebKit/UIProcess/API/glib/WebKitFeaturePrivate.h: Added.
* Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
(webkit_settings_set_feature_enabled):
(webkit_settings_get_feature_enabled):
(webkit_settings_get_all_features):
(webkit_settings_get_experimental_features):
(webkit_settings_get_development_features):
* Source/WebKit/UIProcess/API/glib/WebKitSettings.h.in:
* Source/WebKit/UIProcess/API/glib/webkit.h.in:
* Tools/MiniBrowser/gtk/BrowserSettingsDialog.c:
(featureTreeViewRowActivated):
(cellRendererToggled):
(featureTreeViewRenderEnabledData):
(featureTreeViewRenderStatusData):
(featureTreeViewRenderCategoryData):
(featureTreeViewSearchMatchSubstring):
(createFeatureTreeView):
(settingsTreeViewSearchMatchSubstring):
(browser_settings_dialog_init):
(browserSettingsDialogConstructed):
(browser_settings_dialog_new):
* Tools/MiniBrowser/gtk/main.c:
(findFeature):
(parseFeaturesOptionCallback):
(addSettingsGroupToContext):
* Tools/MiniBrowser/wpe/main.cpp:
(findFeature):
(activate):
(main):
* Tools/Scripts/webkitpy/style/checker.py:
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitSettings.cpp:
(testWebKitFeatures):
(beforeAll):

Canonical link: https://commits.webkit.org/263331@main
